// user/models/user.js
// Part of <https://miracle.systems/p/walkner-paltrack> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.name = 'User';

exports.setUp = (app, mongoose) =>
{
  const userSchema = mongoose.Schema({
    login: {
      type: String,
      trim: true,
      required: true,
      unique: true
    },
    password: {
      type: String,
      trim: true,
      required: true
    },
    email: String,
    privileges: [String],
    partner: {
      type: String,
      ref: 'Partner'
    },
    firstName: String,
    lastName: String
  }, {
    id: false,
    toJSON: {
      transform: function(alarm, ret)
      {
        delete ret.password;

        return ret;
      }
    }
  });

  userSchema.index({lastName: 1});
  userSchema.index({login: 1});
  userSchema.index({email: 1});
  userSchema.index({partner: 1});
  userSchema.index({privileges: 1});

  userSchema.statics.TOPIC_PREFIX = 'users';

  userSchema.statics.customizeLeanObject = function(leanModel)
  {
    if (leanModel)
    {
      delete leanModel.password;
    }

    return leanModel;
  };

  mongoose.model(exports.name, userSchema);
};
