// registry/index.js
// Part of <https://miracle.systems/p/walkner-paltrack> licensed under <CC BY-NC-SA 4.0>

'use strict';

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  partnersId: 'partners',
  palletKindsId: 'palletKinds',
  gdnStoragePath: './gdn',
  wkhtmltopdfUrl: 'http://127.0.0.1/',
  wkhtmltopdfCmd: 'wkhtmltopdf',
  automate: false
};

exports.models = [
  require('./models/dailyBalance'),
  require('./models/gdn'),
  require('./models/grn'),
  require('./models/ob'),
  require('./models/dailyBalance'),
  require('partners/models/partner'),
  require('palletKinds/models/palletKind')
];

exports.userPrivileges = ['REGISTRY:VIEW', 'REGISTRY:MANAGE', 'REGISTRY:CHECK'];

exports.republishTopics = [
  'registry.*.added', 'registry.*.edited', 'registry.*.deleted', 'registry.*.checked.**',
  'balance.daily.**', 'balance.current.*'
];

exports.recordTopics = {
  warning: [
    'registry.*.edited', 'registry.*.deleted'
  ]
};

exports.optionalModules = {
  'mongoose': require('./checker'),
  'mongoose partners': require('./automater'),
  'mongoose partners palletKinds': require('./dailyBalance'),
  'mongoose user express partners palletKinds': require('./routes')
};
